### ISSUE TITLE
Nástroj pro měření vzdálenosti

### ISSUE THREAD

**Issue by mkyral** [osmcz/osmcz#8]
Hodí se pro zjištění vzdálenosti předem naplánované trasy. Takže nestačí jen vzdálenost mezi dvěmi body, ale těch bodů si mohu naklikat kolik bude potřeba.

Výsledná trasa by měla jít sdílet, výškový profil by se také hodil.

**Comment by mkyral** [comment]
Tak jsem si udělal menší průzkum a zjistil, že moduly, které jsou přímo na stránkách Leaflet jsou celkem na nic - moc primitivní.

Zatím nejlepší nástroj, na který jsem zatím narazil, je součástí Geocaching Map Enhancements GreaseMonkey skriptu. https://openuserjs.org/scripts/JRI/Geocaching_Map_Enhancements (MIT licence), github: https://github.com/JRInge/userscripts/tree/master/GeocachingMapEnhancements

Umí naklikat body, smazat bod, přesunout bod (nedynamicky), zobrazit vzdálenost, exportovat GPX. Co neumí, je přidat nový bod na segment. Taky by to chtělo přidat dynamický seznam bodů trasy alá mapy.cz (doufám, že na to nemají patent ;-) )

![gme_measure_tool](https://cloud.githubusercontent.com/assets/155542/12220887/a6997536-b784-11e5-9582-00a8c380c8dd.png)

**Comment by zbycz** [comment]
Přemýšlíš o tom stejně jako já, to se mi líbí :-) pár mých poznámek:
- export by měl být v geojson - šlo by to potom i ukládat na serveru
- můžeme si udělat nějaký panel vlevo, tj. není třeba využívat jen ty ošklivé popupy
- pro kódování do URL by bylo hezké použít [Google Polyline](https://developers.google.com/maps/documentation/utilities/polylineutility) je to otevřená implementace a výrazně zkrátí URL

**Comment by VladimirSlavik** [comment]
Uvažovali jste o [uMap](https://github.com/umap-project/umap)? Sice se neintegruje s "hlavní mapou", ale je to hotové řešení... Ostrá verze viz: http://umap.openstreetmap.fr/

**Comment by zbycz** [comment]
Wow, uMap je moc hezký, to jsem neznal. Asi to bude lepší tlačit spíš v podobě odkazu na stránce využití (?)

**Comment by walley** [comment]
melo by tohle umet routovat? nebo jen merit vzdusnou caru mezi body?

**Comment by mkyral** [comment]
Routování se řeší vedle:  #7 
Tohle by mělo být čistě jen, že si naklikám trasu, kterou chci jít a tohle mi zobrazí vzdálenost mezi jednotlivými body a celkovou vzdálenost. Převážně pro situace, kdy plánuji výlet někam, kde v mapě není cesta, ale vím, že tam ta cesta je (a není zmapovaná), nebo že se tam projít dá (louka).

Ovšem pokud by tam byl mód "přichytávat k cestám", vůbec bych se nezlobil :-D

Plus jako bonus by se mohl zobrazit výškový profil trasy.

**Comment by walley** [comment]
takze https://github.com/Leaflet/Leaflet.Editable plus na to nasmontovane mereni vzdalenosti by mohlo stacit pro zacatek?

**Comment by mkyral** [comment]
Jo. To je pěkné. Bohatě by stačilo kreslení těch čar. Jen je potřeba dodělat měření vzdálenosti. To asi zase vyseparovat jinde.

**Comment by severak** [comment]
Používám knihovnu [MapBBCode](https://github.com/MapBBCode/mapbbcode). Měření délky trasy to umí, vlastně to umí i ten export do geoJson. :-)

Ve výsledku to může být taková odlehčená alternativa pro uMap.

[příklad použití](http://maps.svita.cz/tmyed) (to je tedy klon [mapBBShare](https://github.com/MapBBCode/share.mapbbcode.org) )

**Comment by mkyral** [comment]
Díky, to taky vypadá dobře.

**Comment by zbycz** [comment]
@severak - jo tohle se mi taky velmi zamlouvá. Jen to bude chtít nadesignovat UI, aby to nějak vypadalo. Dodám wireframe.

**Comment by severak** [comment]
Zaznamenal jsem tiket #89 , zkusím tedy s tímto pohnout (ve větvi). Jen upozorňuji, že knihovna MapBB zároveň slouží i jako kreslítko do mapy (takže dubluje #50).

**Comment by zbycz** [comment]
Super, ta MapBB bude velmi dobrý výchozí bod, asi nejlepší co jsem zatím viděl. Ideální varianta je osekat to a udělat z toho Leaflet plugin.

Aby to bylo více user friendly, líbilo by se mi:
1. méně tlačítek, pro začátek zcela stačí vykuchat kreslení čar. Polygony i body bych vynechal (leda pro expertní režim). Uložení pro začátek jen živě do hash URL (formát Google linestring?), v expertním třeba GPX.
2. Krajní body čáry by po kliku neměly mizet, ale pokračovat v kreslení.
3. většinu těch nastavení, pojmenování, barev, simplify, atd ideálně do expertního režimu, pokud vůbec.

Přijde ti to ok?

Jinak aby to šlo dobře začlenit do osmcz-app, bylo by fajn nejprve udělat nějakou jednoduchou variantu a commitnout do svého forku. Bylo by škoda kdybys to dokončil celé a pak jsme zjstili, že se v názorech na to jak to má vypadat zcela rozcházíme :-)

**Comment by severak** [comment]
Měl bych to asi trochu osvětlit: [MapBBcode](http://mapbbcode.org/) je především knihovna sloužící k zobrazování dat (tj. vlastních bodů, čar a polygonů) nad mapou ve webových fórech. 

Potom je to taky tag BB kódu, do kterého se ty data zapisují.

Potom je to taky [MapBBCode Share](http://share.mapbbcode.org/), což je služba postavená nad MapBB kódem sloužící ke sdílení těch vlastních dat.

To, že to umí měřit vzdálenosti  a konvertovat formáty je jen vedlejší efekt (když už si cestu vyznačím, proč jí rovnou nezměřit a neuložit jako GeoJSON).

Já sám tuhle knihovnu [používám](http://maps.svita.cz/ojcyo) (mimo jiné právě na měření vzdálenosti) a věřím, že po zamontování do osmcz by uplatnění určitě našla. Samozřejmě, je nutné ji zamontovat nějak šetrně a to potrvá.

Až budu mít nějakou funkční ukázku nástřelu, dám vědět.

**Comment by zbycz** [comment]
Dík, myslím že to chápu. Bude to chtít tu knihovnu vytěžit a zprovoznit na osmcz. Budu se těšit na nástřel :+1:
